Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm

```diff
@@ -239,30 +240,102 @@ static WKTextExtractionEventListenerTypes eventListenerTypes(OptionSet<TextExtra
 {
     if (RetainPtr textItem = dynamic_objc_cast<WKTextExtractionTextItem>(item)) {
-        TextExtractionFilter::singleton().shouldFilter([textItem content], [textItem, aggregator = Ref { aggregator }](bool shouldFilter) mutable {
-            if (!shouldFilter)
+        TextExtractionFilter::singleton().shouldFilter([textItem content], [
+            enclosingNodeIdentifier = retainPtr(enclosingNodeIdentifier),
+            view = retainPtr(view),
+            textItem,
+            aggregator = Ref { aggregator }
+        ](bool shouldFilter) mutable {
+            if (shouldFilter) {
+                [textItem setContent:@""];
+                [textItem setSelectedRange:NSMakeRange(NSNotFound, 0)];
                 return;
+            }
+
+            RetainPtr nodeIdentifier = [textItem nodeIdentifier] ?: enclosingNodeIdentifier.get();
+            RetainPtr lines = [[textItem content] componentsSeparatedByString:@"\n"];
+
+            auto components = Box<Vector<RetainPtr<NSString>>>::create();
+            components->resizeToFit([lines count]);
+
+            Ref innerAggregator = MainRunLoopCallbackAggregator::create([textItem, aggregator, components] {
+                RetainPtr componentsArray = createNSArray(WTFMove(*components), [](auto component) {
+                    return component.get();
+                });
+                [textItem setContent:[componentsArray componentsJoinedByString:@"\n"]];
+            });
+
+            [lines enumerateObjectsUsingBlock:[&](NSString *substring, NSUInteger index, BOOL*) mutable {
+                static constexpr auto minimumLengthForTextDetection = 100;
+                if (substring.length < minimumLengthForTextDetection) {
+                    components->at(index) = substring;
+                    return;
+                }
 
-            [textItem setContent:@""];
-            [textItem setSelectedRange:NSMakeRange(NSNotFound, 0)];
+                [view _validateText:substring inNode:nodeIdentifier.get() completionHandler:makeBlockPtr([innerAggregator = innerAggregator.copyRef(), components, index](NSString *string) {
+                    components->at(index) = string;
+                }).get()];
+            }];
         });
     }
 
     for (WKTextExtractionItem *child in item.children)
-        filterTextRecursive(child, aggregator);
+        filterTextRecursive(view, child, item.nodeIdentifier ?: enclosingNodeIdentifier, aggregator);
 }
 
-void filterText(WKTextExtractionItem *item, CompletionHandler<void()>&& completion)
+void filterText(WKWebView *view, WKTextExtractionItem *item, CompletionHandler<void()>&& completion)
 {
     Ref aggregator = MainRunLoopCallbackAggregator::create(WTFMove(completion));
-    filterTextRecursive(item, aggregator.get());
+    filterTextRecursive(view, item, nil, aggregator.get());
 }
 
 #endif // ENABLE(TEXT_EXTRACTION_FILTER)
 
+std::optional<double> computeSimilarity(NSString *stringA, NSString *stringB, unsigned minimumLength)
```

Comment:
I wonder if this should use `String` and be moved down into WTF at some point? Not necessary for now.